Implement admin top users

rbennettcw opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Link to Issue
Closes: #6447

## Description of Changes
- Adds route to return list of top 150 users based on activity
- Adds UI component to admin panel to download top users list
- Refactors download logic to be reusable

## Test Plan
- Ensure the user is a super admin
- Go to http://localhost:8080/admin-panel
- At the bottom of the page, click "Generate and Download"– confirm that it downloads a CSV file
- Import CSV into Google Sheets– confirm that it renders correctly

## Deployment Plan
N/A

## Other Considerations
- While downloads are typically handled by the backend via headers,  in this implementation the client fetches JSON and converts to CSV. Not a problem since the list is small.
- JSON to CSV conversion doesn't use a library, so it doesn't catch all edgecases.

jnaviask commented on packages/commonwealth/server/controllers/server_admin_methods/get_top_users.ts:
```diff
@@ -0,0 +1,90 @@
+        GROUP BY
+            user_id, profile_id, profile_name, thread_count, comment_count
+      HAVING
+          NOT ('edgeware' = ANY(array_agg(community_id)) OR 'commonwealth' = ANY(array_agg(community_id)))
```
Need to ask @RhysCommonwealth if this sort of filter is what they wanted, or something more gentle like "admins of either community are removed". Also I believe it's cmn-protocol and not commonwealth as community.

jnaviask replied:
@rbennettcw got some clarification here -- we want to only exclude commonwealth _employees_, not users related to commonwealth. A good proxy there would be checking for super admin on their user object. Or we can remove the filter altogether and the growth team can manually remove, since there's no obvious programmatic way to determine it.

kurtisassad commented on packages/commonwealth/server/controllers/server_admin_methods/get_top_users.ts:
```diff
@@ -0,0 +1,90 @@
+        profile_id,
+        profile_name,
+        array_to_string(addresses_ids, '|') as address_ids,
+        array_to_string(community_ids, '|') as community_ids
```
community_ids looks to be broken, it only returns a single community_id for each user. I will look into this further and respond to this comment if I find the breakage.

rbennettcw replied:
I think it's just that the top users tend to only use a single community.

I increased the limit of the query to show more results and saw a few that had multiple addresses and multiple communities.

rbennettcw replied:
Does that query consider the filtering out of users related to edgeware/cmn-protocol?

kurtisassad replied:
I mean run this top_users query. Find the top user (redacted for privacy). Then plug the user_id into the query in my comment above. Then you get multiple communities as well as multiple addresses of the associated community

rbennettcw replied:
Got it, thanks for digging into it!

kurtisassad replied:
So I can't actually figure out what is wrong with this particular query, but it seems this similar query pulls the correct community/address ids:

```
SELECT 
  p.profile_name AS profile_name,
  COUNT(DISTINCT t.id) AS thread_count,
  COUNT(DISTINCT c.id) AS comment_count,
  COUNT(DISTINCT c.id) + COUNT(DISTINCT t.id) AS total_count,
  ARRAY_AGG(DISTINCT a.community_id ORDER BY a.community_id) AS community_ids,
  ARRAY_AGG(DISTINCT a.id ORDER BY a.id) AS address_ids
FROM "Profiles" AS p
JOIN "Addresses" AS a ON p.id = a.profile_id
LEFT JOIN "Threads" AS t ON a.id = t.address_id
LEFT JOIN "Comments" AS c ON a.id = c.address_id
GROUP BY p.id
ORDER BY TOTAL_COUNT DESC
LIMIT 150;
```

You will still need to add the extra conditions (filter out commonwealth employees), but this might be a better base query to work off for now

rbennettcw commented:
- Updated query to properly return addresses/communities thanks to Kurtis' fix
- Updated query to filter out site admins
- Made column labels human readable
- Added columns for last activity date and days since last active (probably useful to identify inactive users)

jnaviask reviewed (approved):
LGTM
